File: src/lab1/src/lab1.cpp
#include <ros/ros.h>
#include <sensor_msgs/PointCloud2.h>
#include <std_msgs/String.h>
#include <pcl_conversions/pcl_conversions.h>
#include <pcl/point_cloud.h>
#include <pcl/point_types.h>
#include <pcl/filters/passthrough.h>

// Note: global variables like this are bad practice, but is done here for simplicity.
// A more elegant solution would be to write your ROS node as a class, and have the
// publishers and subscribers be private variables inside said class.
ros::Publisher pcl_pub;

void pointcloud_callback(const sensor_msgs::PointCloud2ConstPtr & input_msg)
{
  pcl::PointCloud<pcl::PointXYZI>::Ptr input_pcl(new pcl::PointCloud<pcl::PointXYZI>);
  pcl::fromROSMsg(*input_msg, *input_pcl);
  // This code runs fast, but can be a bit difficult to read if you aren't
  // familiar with libPCL. An loop that performs the same operation but is slower
  // has been provided for your reference
  /*
  for(auto it = input_pcl->begin(); it<input_pcl->end(); it++)
  {
    if(it->x < 0 or it->x > 100)
    {
      input_pcl->erase(it);
      --it;
    }
  }
  */
  pcl::PassThrough<pcl::PointXYZI> filter;
  pcl::PointCloud<pcl::PointXYZI>::Ptr filtered(new pcl::PointCloud<pcl::PointXYZI>);
  filter.setInputCloud(input_pcl);
  filter.setFilterFieldName("x");
  filter.setFilterLimits(0.0,100.0);
  filter.filter(*filtered);
  sensor_msgs::PointCloud2 output_msg;
  pcl::toROSMsg(*filtered, output_msg);

  output_msg.header = input_msg->header;
  pcl_pub.publish(output_msg);
}

int main(int argc, char * argv[])
{
  ros::init(argc, argv, "lab1_node");
  ros::NodeHandle nh;
  pcl_pub = nh.advertise<sensor_msgs::PointCloud2>("points_modified", 2, true);
  ros::Subscriber pcl_sub = nh.subscribe("points_raw", 2, pointcloud_callback);
  ros::spin();
  return 0;
}

File: src/lab1/py/lab1.py
#!/usr/bin/env python

import rospy
from sensor_msgs.msg import PointCloud2, PointField
import sensor_msgs.point_cloud2 as pc2
from std_msgs.msg import String

class lab_1_node():
  def __init__(self):
    self.pcl_sub = rospy.Subscriber('/points_raw', PointCloud2, self.pointcloud_callback)
    self.pcl_pub = rospy.Publisher('/points_modified', PointCloud2, queue_size=2)

  def pointcloud_callback(self, msg):
    points = []
    field_names = ('x', 'y', 'z', 'intensity')
    for p in pc2.read_points(msg, field_names = field_names, skip_nans=True):
      if p[0] >= 0:
        points.append(p)
    fields = [PointField('x', 0, PointField.FLOAT32,1),
              PointField('y', 4, PointField.FLOAT32,1),
              PointField('z', 8, PointField.FLOAT32,1),
              PointField('intensity', 12, PointField.UINT32,1),]
    output = pc2.create_cloud(msg.header, fields, points)
    self.pcl_pub.publish(output);

if __name__ == '__main__':
  try:
    rospy.init_node('lab1_py_node', anonymous=True)
    lab_1_node()
    rospy.spin()
  except RuntimeError as e:
    pass
